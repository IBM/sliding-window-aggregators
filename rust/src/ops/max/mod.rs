use std::cmp::{Ord, PartialEq};
use std::marker::PhantomData;

use num_traits::cast::{ToPrimitive, NumCast};

use alga::general::AbstractMagma;
use alga::general::AbstractMonoid;
use alga::general::AbstractSemigroup;
use alga::general::Identity;
use alga::general::Operator;

use super::AggregateOperator;
use super::AggregateMonoid;

// We need a generic way to talk about a type's minimum 
// value
pub trait Min {
    fn min() -> Self;
}

impl Min for i32 {
    fn min() -> i32 {
        i32::MIN
    }
}

/// Binary operator for maximum.
/// Has the following properties:
/// * Associativity
/// * Commutativity
#[derive(Copy, Clone)]
pub struct Max<In, Out>
where
    In: Ord + Min + Copy,
    Out: Ord + Min + Copy,
{
    in_type: PhantomData<In>,
    out_type: PhantomData<Out>
}

// We need this because of the orphan rules for trait implemenations. See 
// `rustc --explain E0210`.
#[derive(Copy, Clone, Eq, PartialEq)]
pub struct MaxPartial<T> {
    val: T
}

impl<In, Out> AggregateMonoid<Max<In, Out>> for Max<In, Out>
where
    In: Ord + Min + ToPrimitive + Copy,
    Out: Ord + Min + NumCast + Copy,
{
    type Partial = MaxPartial<Out>;

    fn lift(v: In) -> Self::Partial {
        Self::Partial{
            val: NumCast::from(v).unwrap()
        }
    }
    fn lower(part: &Self::Partial) -> Out {
        part.val
    }
}

impl<In_, Out_> AggregateOperator for Max<In_, Out_>
where
    In_: Ord + Min + ToPrimitive + Copy,
    Out_: Ord + Min + NumCast + Copy,
{
    type In = In_;
    type Out = Out_;
}

impl<T: Ord + Min + NumCast + ToPrimitive + Copy> Max<T, T> {
    pub fn name() -> &'static str {
        "max"
    }
}

impl<In, Out> Operator for Max<In, Out> 
where
    In: Ord + Min + ToPrimitive + Copy,
    Out: Ord + Min + NumCast + Copy,
{
    fn operator_token() -> Self {
        Self {
            in_type: PhantomData,
            out_type: PhantomData
        }
    }
}

impl<In, Out> Identity<Max<In, Out>> for MaxPartial<Out>
where
    In: Ord + Min + ToPrimitive + Copy,
    Out: Ord + Min + NumCast + Copy,
{
    fn identity() -> Self {
        Self {
            val: Min::min()
        }
    }
}

impl<In, Out> AbstractMagma<Max<In, Out>> for MaxPartial<Out>
where
    In: Ord + Min + ToPrimitive + Copy,
    Out: Ord + Min + NumCast + Copy,
{
    fn operate(&self, other: &Self) -> Self {
        if self.val > other.val {
            Self {
                val: self.val
            }
        } else {
            Self {
                val: other.val
            }
        }
    }
}

impl<In, Out> AbstractSemigroup<Max<In, Out>> for MaxPartial<Out>
where
    In: Ord + Min + ToPrimitive + Copy,
    Out: Ord + Min + NumCast + Copy,
{}

impl<In, Out> AbstractMonoid<Max<In, Out>> for MaxPartial<Out>
where
    In: Ord + Min + ToPrimitive + Copy,
    Out: Ord + Min + NumCast + Copy,
{}


