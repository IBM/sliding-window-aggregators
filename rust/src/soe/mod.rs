use crate::FifoWindow;
use alga::general::AbstractGroup;
use alga::general::Operator;
use std::collections::VecDeque;
use std::marker::PhantomData;

#[derive(Clone)]
pub struct SoE<Value, BinOp>
where
    Value: AbstractGroup<BinOp> + Clone,
    BinOp: Operator,
{
    stack: VecDeque<Value>,
    agg: Value,
    op: PhantomData<BinOp>,
}

impl<Value, BinOp> FifoWindow<Value, BinOp> for SoE<Value, BinOp>
where
    Value: AbstractGroup<BinOp> + Clone,
    BinOp: Operator,
{
    fn new() -> Self {
        Self {
            stack: VecDeque::new(),
            agg: Value::identity(),
            op: PhantomData,
        }
    }
    fn push(&mut self, v: Value) {
        self.agg = self.agg.operate(&v);
        self.stack.push_back(v);
    }
    fn pop(&mut self) {
        if let Some(top) = self.stack.pop_front() {
            self.agg = self.agg.operate(&top.two_sided_inverse());
        }
    }
    fn query(&self) -> Value {
        self.agg.clone()
    }
    fn len(&self) -> usize {
        self.stack.len()
    }
}
