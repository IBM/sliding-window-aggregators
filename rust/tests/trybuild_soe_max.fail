use swag::soe::*;
use swag::ops::max::Max;

/// Should not compile since Max is not invertible
fn main() {
    SoE::<Max<i32, i32>>::new();
}
