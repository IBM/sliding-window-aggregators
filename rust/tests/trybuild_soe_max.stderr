error[E0599]: no function or associated item named `new` found for struct `swag::soe::SoE<common::Int, common::Max>` in the current scope
 --> $DIR/trybuild_soe_max.fail:7:22
  |
7 |     SoE::<Int, Max>::new();
  |                      ^^^ function or associated item not found in `swag::soe::SoE<common::Int, common::Max>`

error[E0277]: the trait bound `common::Int: alga::general::one_operator::AbstractGroup<common::Max>` is not satisfied
  --> $DIR/trybuild_soe_max.fail:7:5
   |
7  |     SoE::<Int, Max>::new();
   |     ^^^^^^^^^^^^^^^^^^^^ the trait `alga::general::one_operator::AbstractGroup<common::Max>` is not implemented for `common::Int`
   |
  ::: $WORKSPACE/src/soe/mod.rs:10:12
   |
10 |     Value: AbstractGroup<BinOp> + Clone,
   |            -------------------- required by this bound in `swag::soe::SoE`
   |
   = help: the following implementations were found:
             <common::Int as alga::general::one_operator::AbstractGroup<common::Sum>>
